miio/ihcooker.py

```diff
@@ -0,0 +1,746 @@
+
+MODEL_EG1 = "chunmi.ihcooker.eg1"
+MODEL_EXP1 = "chunmi.ihcooker.exp1"
+MODEL_FW = "chunmi.ihcooker.chefnic"
+MODEL_HK1 = "chunmi.ihcooker.hk1"
+MODEL_KOREA1 = "chunmi.ihcooker.korea1"
+MODEL_TW1 = "chunmi.ihcooker.tw1"
+MODEL_V1 = "chunmi.ihcooker.v1"
+
+MODEL_VERSION1 = [MODEL_V1, MODEL_FW, MODEL_HK1, MODEL_TW1]
+MODEL_VERSION2 = [MODEL_EG1, MODEL_EXP1, MODEL_KOREA1]
+DEVICE_ID = {
+    MODEL_EG1: 2,  # TODO: TBD
+    MODEL_EXP1: 4,
+    MODEL_FW: 1,  # TODO: TBD
+    MODEL_HK1: 1,  # TODO: TBD
+    MODEL_KOREA1: 2,  # TODO: TBD
+    MODEL_TW1: 1,  # TODO: TBD
+    MODEL_V1: 1,
+}
+DEVICE_PREFIX = {k: "03%02d" % v for k, v in DEVICE_ID.items()}
+
+VERSION_1 = 1
+VERSION_2 = 2
+HARDWARE_VERSIONS = {
+    1: VERSION_1,
+    2: VERSION_2,  # Guess.
+    3: VERSION_2,  # Guess.
+    4: VERSION_2,
+    5: VERSION_2,  # Guess.
+}
+
+
+class IHCookerException(DeviceException):
+    pass
+
+
+class StageMode(enum.Enum):
+    TimerMode = 8
+    TemperatureMode = 2
+    FireMode = 0
+
+
+class OperationMode(enum.Enum):
+    Error = "error"
+    Finish = "finish"
+    Offline = "offline"
+    Pause = "pause"
+    Pause_time = "pause_time"
+    Precook = "precook"
+    Running = "running"
+    Set_01 = "set01"
+    Set_02 = "set02"
+    Set_03 = "set03"
+    Shutdown = "shutdown"
+    Timing = "timing"
+    Waiting = "waiting"
+
+
+def calcrc(data, n):
+    """
+     public static int calcrc(byte[] bArr, int i) {
+        int i2 = 0;
+        int i3 = 0;
+        while (true) {
+            i--;
+            if (i < 0) {
+                return 65535 & i2;
+            }
+            int i4 = i3 + 1;
+            int i5 = i2 ^ (bArr[i3] << 8);
+            for (int i6 = 0; i6 < 8; i6++) {
+                i5 = (32768 & i5) != 0 ? (i5 << 1) ^ 4129 : i5 << 1;
+            }
+            i2 = i5;
+            i3 = i4;
+        }
+    }
+    """
+    i2 = 0
+    i3 = 0
+    while True:
+        n -= 1
+        if n < 0:
+            return 65535 & i2
+
+        i4 = i3 + 1
+        i5 = i2 ^ (data[i3] << 8)
+        for i6 in range(8):
+            i5 = ((i5 << 1) ^ 4129) if (32768 & i5) != 0 else (i5 << 1)
+
+        i2 = i5
+        i3 = i4
+
+
+class CookProfile:
+    def __init__(self, model, profile=None):
+        self.model = model
+        """Initialize a cooking profile from an existing one, or a new one."""
+        if profile is not None:
+            self.data = bytearray.fromhex(profile)
+        else:
+            self.data = bytearray([0] * 179)
+            # Initialize recipe phases.
+            if self.is_v1:
+                offset = 38
+            else:
+                offset = 48
+
+            for i in range(offset, 16 * 8, 8):
+                self.data[i + 0] = 0
+                self.data[i + 1] = 0
+                self.data[i + 2] = 0
+                self.data[i + 3] = 249
+                self.data[i + 4] = 229
+                self.data[i + 5] = 0
+                self.data[i + 6] = 20
+                self.data[i + 7] = 20
```

Comment:
If you know (at least parts) of the structure, it'd be great to have this in `construct` structs as used for the miio protocol itself for maintainability, see https://construct.readthedocs.io/en/latest/intro.html

Reply:
I wonder if the format is similar in any way to chunmi.cooker.normalX models. Did you happen to document any of your reverse engineering efforts?

Reply:
from my knowing :
`self.data[i + 0] = 0 # mode: heat_until_temp, heat_until_timer, continuous_output.
self.data[i + 1] = 0 # hours: how long this phase lasts in minutes.
self.data[i + 2] = 0 #mins: how long this phase lasts in minutes.
self.data[i + 3] = 249 #thresh: temperature threshold for moving to next phase, in celcius.
self.data[i + 4] = 229 #temp: target temperature to heat to, in celsius.
self.data[i + 5] = 0 #fire: power output between [0,99]

#set continuous or pulse fire
#same value for 2 = continous heat
#different value = pulse heat, usefull for low heat adjustment
#0 = no heat
self.data[i + 6] = 20 #fire off time
self.data[i + 7] = 20 #fire on time`

Reply:
@coolibry Thanks, will tinker and see if it fits my cooker.

Reply:
@rytilahti : Using Construct here makes sense I think. I propose to leave this open for follow-up pull request though.